### ISSUE TITLE
bug: kcp panics when trying delete bound CRD and a corresponding apibinding exists 

### ISSUE THREAD

**Issue by kylape** [kcp-dev/kcp#2304]
### Describe the bug

The `apis.kcp.dev/identity` annotation on bound CRDs is added by some handler for every custom resource request.  It's used to add the `APIExport` identity hash to the etcd key prefix.

When a CRD is being deleted, the handler that adds this annotation is not executed, and the same code path that [checks for `apis.kcp.dev/identity`](https://github.com/kcp-dev/kubernetes/blob/38c73163e76618f5ba6aef37baac2c86954e8652/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go#L52-L55
) is executed when trying to delete instances of the CRD.

```E1103 14:24:57.457033 3401974 runtime.go:79] Observed a panic: &errors.errorString{s:"missing 'apis.kcp.dev/identity' annotation on CRD system:bound-crds|36973e4e-566d-4e58-9c7b-0901aae87636 for tlsroutes.gateway.networking.k8s.io"} (missing 'apis.kcp.dev/identity' annotation on CRD system:bound-crds|36973e4e-566d-4e58-9c7b-0901aae87636 for tlsroutes.gateway.networking.k8s.io)
goroutine 1885 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x42bf1c0?, 0xc01516aaa0})
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00bdaaa10?})
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x75
panic({0x42bf1c0, 0xc01516aaa0})
	/usr/local/go/src/runtime/panic.go:838 +0x207
k8s.io/apiextensions-apiserver/pkg/registry/customresource.newStores({{_, _}, {_, _}}, {{0xc0038e3000, 0x19}, {0xc00c7eaf00, 0x8}, {0xc00c7eba68, 0x8}}, ...)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go:111 +0x6cc
k8s.io/apiextensions-apiserver/pkg/registry/customresource.newREST({{_, _}, {_, _}}, {{0xc0038e3000, 0x19}, {0xc00c7eaf00, 0x8}, {0xc00c7eba68, 0x8}}, ...)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go:126 +0x10f
k8s.io/apiextensions-apiserver/pkg/registry/customresource.NewStorageWithCustomStore({{_, _}, {_, _}}, {{0xc0038e3000, 0x19}, {0xc00c7eaf00, 0x8}, {0xc00c7eba68, 0x8}}, ...)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go:50 +0x10e
k8s.io/apiextensions-apiserver/pkg/registry/customresource.NewStorage(...)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go:46
k8s.io/apiextensions-apiserver/pkg/apiserver.(*crdHandler).getOrCreateServingInfoFor(0xc000630700, 0xc010b9f900)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:947 +0x1ac6
k8s.io/apiextensions-apiserver/pkg/apiserver.(*crdHandler).GetCustomResourceListerCollectionDeleter(0xc015a28400?, 0x43ad300?)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:732 +0x19
k8s.io/apiextensions-apiserver/pkg/controller/finalizer.(*CRDFinalizer).deleteInstances(0xc0119dd4e0?, 0xc010b9f900)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go:183 +0x67
k8s.io/apiextensions-apiserver/pkg/controller/finalizer.(*CRDFinalizer).sync(0xc001849590, {0xc0120bb740?, 0xc00008c9b0?})
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go:153 +0x807
k8s.io/apiextensions-apiserver/pkg/controller/finalizer.(*CRDFinalizer).processNextWorkItem(0xc001849590)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go:297 +0x106
k8s.io/apiextensions-apiserver/pkg/controller/finalizer.(*CRDFinalizer).runWorker(0xc0039c56a0?)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go:285 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x1?)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x413db80?, {0x4f445e0, 0xc003d74450}, 0x1, 0xc002800420)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x1?, 0x16ee6a5?)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(0x0?, 0xc002800420?, 0x0?)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x25
created by k8s.io/apiextensions-apiserver/pkg/controller/finalizer.(*CRDFinalizer).Run
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go:278 +0x1e5
panic: missing 'apis.kcp.dev/identity' annotation on CRD system:bound-crds|36973e4e-566d-4e58-9c7b-0901aae87636 for tlsroutes.gateway.networking.k8s.io [recovered]
	panic: missing 'apis.kcp.dev/identity' annotation on CRD system:bound-crds|36973e4e-566d-4e58-9c7b-0901aae87636 for tlsroutes.gateway.networking.k8s.io

goroutine 1885 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00bdaaa10?})
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xd8
panic({0x42bf1c0, 0xc01516aaa0})
	/usr/local/go/src/runtime/panic.go:838 +0x207
k8s.io/apiextensions-apiserver/pkg/registry/customresource.newStores({{_, _}, {_, _}}, {{0xc0038e3000, 0x19}, {0xc00c7eaf00, 0x8}, {0xc00c7eba68, 0x8}}, ...)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go:111 +0x6cc
k8s.io/apiextensions-apiserver/pkg/registry/customresource.newREST({{_, _}, {_, _}}, {{0xc0038e3000, 0x19}, {0xc00c7eaf00, 0x8}, {0xc00c7eba68, 0x8}}, ...)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go:126 +0x10f
k8s.io/apiextensions-apiserver/pkg/registry/customresource.NewStorageWithCustomStore({{_, _}, {_, _}}, {{0xc0038e3000, 0x19}, {0xc00c7eaf00, 0x8}, {0xc00c7eba68, 0x8}}, ...)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go:50 +0x10e
k8s.io/apiextensions-apiserver/pkg/registry/customresource.NewStorage(...)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go:46
k8s.io/apiextensions-apiserver/pkg/apiserver.(*crdHandler).getOrCreateServingInfoFor(0xc000630700, 0xc010b9f900)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:947 +0x1ac6
k8s.io/apiextensions-apiserver/pkg/apiserver.(*crdHandler).GetCustomResourceListerCollectionDeleter(0xc015a28400?, 0x43ad300?)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:732 +0x19
k8s.io/apiextensions-apiserver/pkg/controller/finalizer.(*CRDFinalizer).deleteInstances(0xc0119dd4e0?, 0xc010b9f900)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go:183 +0x67
k8s.io/apiextensions-apiserver/pkg/controller/finalizer.(*CRDFinalizer).sync(0xc001849590, {0xc0120bb740?, 0xc00008c9b0?})
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go:153 +0x807
k8s.io/apiextensions-apiserver/pkg/controller/finalizer.(*CRDFinalizer).processNextWorkItem(0xc001849590)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go:297 +0x106
k8s.io/apiextensions-apiserver/pkg/controller/finalizer.(*CRDFinalizer).runWorker(0xc0039c56a0?)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go:285 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x1?)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x413db80?, {0x4f445e0, 0xc003d74450}, 0x1, 0xc002800420)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x1?, 0x16ee6a5?)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(0x0?, 0xc002800420?, 0x0?)
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x25
created by k8s.io/apiextensions-apiserver/pkg/controller/finalizer.(*CRDFinalizer).Run
	/home/klape/src/k8s/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/finalizer/crd_finalizer.go:278 +0x1e5
exit status 2
```

### Steps To Reproduce

1. Create a valid `APIExport`
2. Create an `APIBinding` for the export
3. Get the UID of one of the `APIResourceSchemas`.  Look at `apibinding.status.boundResources[0].schema.UID`.
4. Delete the corresponding CRD in the `system:bound-crds` workspace.

### Expected Behaviour

Some kind of error that does not allow the CRD to be deleted with existing `APIBindings` that use it.

### Additional Context

This problem was encountered when trying to solve https://github.com/kcp-dev/kcp/issues/2300

**Comment by csams** [comment]
Panic reference: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go#L71

**Comment by kylape** [comment]
Fixed by #2298
